README.md
# haensl.github.io

Source for [https://haensl.github.io](https://haensl.github.io)

![build status](https://gitlab.com/hpd/haensl.github.io.src/badges/master/pipeline.svg)

## Quick Start

```
npm i
npm start
```

### Employed technologies

- [AMP](https://www.ampproject.org)
- [Docker](https://www.docker.com/)
- [Koa](https://koajs.com/)
- [Gulp](https://gulpjs.com/)
- [Mustache](https://mustache.github.io/)
- [Node.js](https://nodejs.org)

### [CHANGELOG](CHANGELOG.md)

[Licensed](LICENSE) under GPLv3.0


CHANGELOG.md
### 3.2.4
* [#300: Update dependencies.](https://github.com/haensl/haensl.github.io.src/issues/300)

### 3.2.3
* [#297: Update dependencies.](https://github.com/haensl/haensl.github.io.src/issues/297)

### 3.2.2
* [#295: Update dependencies.](https://github.com/haensl/haensl.github.io.src/issues/295)

### 3.2.1
* [#293: Update dependencies.](https://github.com/haensl/haensl.github.io.src/issues/293)

### 3.2.0
* [#240: Add moonlight to projects.](https://github.com/haensl/haensl.github.io.src/issues/240)

### 3.1.0
* [#237: Add json-parser to projects.](https://github.com/haensl/haensl.github.io.src/issues/237)

### 3.0.19
* [#289: Update dependencies.](https://github.com/haensl/haensl.github.io.src/issues/289)
* Update react-component-console NPM link.

### 3.0.18
* [#287: Update project links.](https://github.com/haensl/haensl.github.io.src/issues/287)
* Update dependencies.

### 3.0.17
* [#284: Fix HTML entities on CV page.](https://github.com/haensl/haensl.github.io.src/issues/284)
* Update dependencies.

### 3.0.16
* [#265: Fix hpdietz.de Google Analytics configuration.](https://github.com/haensl/haensl.github.io.src/issues/265)
* Update dependencies.

### 3.0.15
* [#280: Fix page title.](https://github.com/haensl/haensl.github.io.src/issues/280)

### 3.0.14
* [#280: Fix page title.](https://github.com/haensl/haensl.github.io.src/issues/280)
* Update dependencies.

### 3.0.13
* [#277: Fix german HTML entities.](https://github.com/haensl/haensl.github.io.src/issues/277)

### 3.0.12
* [#277: Fix german HTML entities.](https://github.com/haensl/haensl.github.io.src/issues/277)
* Update dependencies.

### 3.0.11
* [#275: Fix HTML entities.](github.com/haensl/haensl.github.io.src/issues/275)

### 3.0.10
* [#273: Fix software mustache template.](github.com/haensl/haensl.github.io.src/issues/273)

### 3.0.9
* [#271: Update haensl.github.io build task.](github.com/haensl/haensl.github.io.src/issues/271)

### 3.0.8
* [#269: Update dependencies.](github.com/haensl/haensl.github.io.src/issues/269)

### 3.0.6
* [#265: Fix hpdietz.de Google Analytics configuration.](https://github.com/haensl/haensl.github.io.src/issues/265)
* Update dependencies.

### 3.0.5
* [#260: Fix mobile UI.](https://github.com/haensl/haensl.github.io.src/issues/260)

### 3.0.4
* [#260: Fix mobile UI.](https://github.com/haensl/haensl.github.io.src/issues/260)

### 3.0.3
* [#257: Bump ini from 1.3.5 to 1.3.8 via dependabot.](https://github.com/haensl/haensl.github.io.src/pull/257))
### 3.0.7
* [#267: Update CVs.](https://github.com/haensl/haensl.github.io.src/issues/267)

### 3.0.6
* [#265: Fix hpdietz.de Google Analytics configuration.](https://github.com/haensl/haensl.github.io.src/issues/265)
* Update dependencies.

### 3.0.5
* [#260: Fix mobile UI.](https://github.com/haensl/haensl.github.io.src/issues/260)

### 3.0.4
* [#260: Fix mobile UI.](https://github.com/haensl/haensl.github.io.src/issues/260)

### 3.0.3
* [#257: Bump ini from 1.3.5 to 1.3.8 via dependabot.](https://github.com/haensl/haensl.github.io.src/pull/257)

### 3.0.2
* [#260: Fix mobile UI.](https://github.com/haensl/haensl.github.io.src/issues/260)
* Update dependencies.

### 3.0.1
* [#258: Fix german Docker image.](https://github.com/haensl/haensl.github.io.src/issues/258)

### 3.0.0
* Localize application:
  * Add new domain: https://hpdietz.de for german visitors.
  * Use existing domains for english/international visitors.
  * Refactor build process.

### 2.62.0
* [#252: Add "hire me" feature.](https://github.com/haensl/haensl.github.io.src/issues/252)
* Add services page.
* Add `@haensl/json-transform-stream` to projects.
* Change author `@id` to `https://hpdietz.com`.

### 2.61.2
* [#250: Remove obsolete pull when building Docker image.](https://github.com/haensl/haensl.github.io.src/issues/250)

### 2.61.1
* [#248: Invert container name.](https://github.com/haensl/haensl.github.io.src/issues/248)
* Use `npm ci` command for container.
* Use buildkit when building Docker image.

### 2.61.0
* [#245: Remove stack.](https://github.com/haensl/haensl.github.io.src/issues/245)
* Replace publications & talks with resources.
* Rename projects to software.
* Add teaching to resources.

### 2.60.0
* [#244: Add @haensl/services to projects.](https://github.com/haensl/haensl.github.io.src/issues/244)
* Remove build badges.
* Increase project name font-size.
* Update dependencies.

### 2.59.0
* [#241: Add @haensl/hooks to projects.](https://github.com/haensl/haensl.github.io.src/issues/241)

### 2.58.0
* [#238: Add experteer to CV.](https://github.com/haensl/haensl.github.io.src/issues/238)
* Update dependencies.

### 2.57.0
* [#235: Add build badge.](https://github.com/haensl/haensl.github.io.src/issues/235)

### 2.56.1
* [#229: Update used technologies in readme.](https://github.com/haensl/haensl.github.io.src/issues/229)

### 2.56.0
* [#231: Add pfs to projects.](https://github.com/haensl/haensl.github.io.src/issues/231)

### 2.55.1
* [#230: react-component-console: Replace Travis badge with CircleCI badge.](https://github.com/haensl/haensl.github.io.src/issues/230)
* Update dependencies.

### 2.55.0
* [#227: Add google-optimize-service to projects.](https://github.com/haensl/haensl.github.io.src/issues/227)

### 2.54.0
* [#225: Add link to projects to about page text.](https://github.com/haensl/haensl.github.io.src/issues/225)

### 2.53.2
* [#223: Update copyright year.](https://github.com/haensl/haensl.github.io.src/issues/223)
* Update dependencies.

### 2.53.1
* [#221: Mobile style fixes.](https://github.com/haensl/haensl.github.io.src/issues/221)

### 2.53.0
* [#142: Migrate to koa.](https://github.com/haensl/haensl.github.io.src/issues/142)

### 2.52.2
* [#218: Fix mobile burger menu.](https://github.com/haensl/haensl.github.io.src/issues/218)

### 2.52.1
* [#216: Change email to gmail.](https://github.com/haensl/haensl.github.io.src/issues/216)
* Update dependencies.

### 2.52.0
* Update CV.

### 2.51.0

src/app.js
const Koa = require('koa');
const serve = require('koa-static');
const compress = require('koa-compress');
const etag = require('koa-etag');
const conditional = require('koa-conditional-get');
const send = require('koa-send');
const gracefulShutdown = require('http-graceful-shutdown');
const hostname = require('os').hostname();
const path = require('path');
const Z_SYNC_FLUSH = require('zlib').Z_SYNC_FLUSH;
const log = require('@haensl/log');
const {
  name,
  version
} = require('./package');

const statuscodes = {
  NOT_FOUND: 404,
  INTERNAL_SERVER_ERROR: 500
};

const publicDir = process.env.PUBLIC_DIR ||
  path.resolve(__dirname, './public/');

const start = async (port) => {
  const server = new Koa();
  server
    .use(async (ctx, next) => {
      try {
        await next();
        const status = ctx.status || statuscodes.NOT_FOUND;
        if (status === statuscodes.NOT_FOUND) {
          ctx.throw(statuscodes.NOT_FOUND);
        }
      } catch (err) {
        ctx.status = err.status || statuscodes.INTERNAL_SERVER_ERROR;
        let errorPage;
        switch (ctx.status) {
          case 404:
            log.debug('404: Not Found', ctx.request);
            ctx.redirect('/404');
            break;
          default:
            log.error('500: Internal Server Error', err, 'Request', ctx.request);
            ctx.redirect('/500');
            break;
        }
      }
    })
    .use(compress({
      threshold: process.env.COMPRESSION_THRESHOLD || 0,
      flush: Z_SYNC_FLUSH
    }))
    .use(conditional())
    .use(etag())
    .use(serve(publicDir, {
      maxage: process.env.PUBLIC_CACHE_MAXAGE || 0
    }));

  await server.listen(port);

  log.info(`${name}@${version} listening on http://${hostname}:${port}`);
  log.debug(`serving ${publicDir}`);

  log.debug('attaching shutdown handler');
  gracefulShutdown(
    server,
    {
      development: process.env.NODE_ENV !== 'production',
      finally: () => {
        log.info(`${name}@${version} shutting down gracefully.`);
        process.exit(0);
      }
    }
  );

  server.on('error', (error) => {
    log.error('Uncaught Koa error', error);
  });
};

start(process.env.PORT || 8080)
  .catch((error) => {
    log.error(`Failing to start ${name}@${version}`, error);
    process.exit(1);
  });


gulpfile.js
const gulp = require('gulp');
const pfs = require('@haensl/pfs');
const path = require('path');
const $ = require('gulp-load-plugins')();
const del = require('del');
const browserSync = require('browser-sync');
const log = require('@haensl/log');
const pkg = require('./package')

const VERSION = pkg.version;
const DIR_DIST = 'dist';
const DIR_SRC = 'src';
const DIR_SRC_ASSETS = `${DIR_SRC}/assets`;
const DIR_SRC_SEOFILES=`${DIR_SRC}/seofiles`;
const DIR_TEMPLATES = `${DIR_SRC}/templates`;
const DIR_ARTWORK = `${DIR_SRC}/artwork`;
const DIRECTIVE_CSS_INCLUDE = '<!-- INCLUDE_CSS -->';
const DIRECTIVE_INCLUCE_VERSION = '<!-- INCLUDE_VERSION -->';
const DIRECTIVE_INCLUDE_REVISED = '<!-- INCLUDE_REVISED -->';
const sites = require(`./${DIR_TEMPLATES}/sites`);
const DATE_NOW_ISO = (new Date()).toISOString();
const OPTS_HTMLMIN = {
  collapseWhitespace: true,
  removeComments: true,
  minifyJS: true
};

const FILES_SERVER = [
  'package.json',
  'package-lock.json',
  `${DIR_SRC}/app.js`
];

const dirExists = (dir) =>
  (pfs.existsSync(dir)
    && pfs.statSync(dir).isDirectory());

const createDir = (dir) =>
  dirExists(dir)
    || pfs.mkdirSync(dir);

const validateAMP = () =>
  gulp.src(`${DIR_DIST}/**/index.html`)
    .pipe($.amphtmlValidator.validate());
    //.pipe($.amphtmlValidator.format());
    //.pipe($.amphtmlValidator.failAfterError());

gulp.task('ensureDistDirExists', (done) => {
  createDir(DIR_DIST)
  done();
});

gulp.task('ensureDistDirsExists',
  gulp.series(
    'ensureDistDirExists',
    (done) => {
      for (const domain of pkg.domains) {
        createDir(path.join(DIR_DIST, `/${domain.domain}`));
        if (domain.includeServer) {
          createDir(path.join(DIR_DIST, `/${domain.domain}/public`));
          createDir(path.join(DIR_DIST, `/${domain.domain}/public/assets`));
        }
      }

      done();
    }
  )
);

gulp.task('clean:seofiles',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      [
        `${DIR_DIST}/**/google*+.html`,
        `${DIR_DIST}/**/robots.txt`
      ],
      { force: true }
    )
  )
);

gulp.task('clean:sitemap',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      `${DIR_DIST}/**/sitemap.xml`,
      { force: true }
    )
  )
);

gulp.task('clean:serverfiles',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      FILES_SERVER.map((file) => `${DIR_DIST}/**/${file}`),
      { force: true }
    )
  )
);

gulp.task('clean:css',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      [ `${DIR_DIST}/**/*.css` ],
      { force: true }
    )
  )
);

gulp.task('clean:html',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      [`${DIR_DIST}/**/*.html`],
      { force: true }
    )
  )
);

gulp.task('clean:assets',
  gulp.series(
    'ensureDistDirsExists',
    () => del(
      [ `${DIR_DIST}/**/assets/*` ],
      { force: true }
    )
  )
);

gulp.task('seofiles',
  gulp.series(
    'clean:seofiles',
    () =>
      Promise.all(
        pkg.domains.map((domain) => new Promise((resolve, reject) => {
            gulp.src(`${DIR_SRC_SEOFILES}/${domain}/*`)
              .pipe(gulp.dest(`${DIR_DIST}/${domain}/`))
              .on('end', resolve)
              .on('error', reject);
          })
        )
      )
  )
);

gulp.task('server',
  gulp.series(
    'clean:serverfiles',
    () => Promise.all(
      pkg.domains
        .filter((domain) => domain.includeServer)
        .map((domain) => new Promise((resolve, reject) =>
          gulp.src(FILES_SERVER)
            .pipe(gulp.dest(`${DIR_DIST}/${domain.domain}/`))
            .on('end', resolve)
            .on('error', reject))
        )
    )
  )
);

gulp.task('css',
  gulp.series(
    'clean:css',
    () => new Promise((resolve, reject) =>
        gulp.src(`${DIR_SRC}/*.css`)
          .pipe($.postcss([require('autoprefixer')()]))
          .pipe($.cleanCss())
          .pipe($.rename((path) => {
            path.basename = `${path.basename}.tmp`
            return path;
          }))
          .pipe(gulp.dest(DIR_DIST))
          .on('end', resolve)
          .on('error', reject)
    )
  )
);

gulp.task('sitemap',
  gulp.series(
    'clean:sitemap',
    () => Promise.all(
      pkg.domains.map((domain) => new Promise((resolve, reject) => {
        const now = (new Date()).toISOString();
        gulp.src(`${DIR_TEMPLATES}/sitemap/template.mustache`)
          .pipe($.mustache({
            domain: domain.domain,
            modified: now
          }))
          .pipe($.rename('sitemap.xml'))
          .pipe(gulp.dest(`${DIR_DIST}/${domain.domain}/${domain.includeServer ? 'public' : ''}`))
          .on('end', resolve)
          .on('error', reject)
      }))
    )
  )
);

gulp.task(
  'templates',
  gulp.series(
    'clean:html',
    async (done) => {
      const components = [
        'footer-menu',
        'language-menu',
        'menu',
        'social-nav',
        'social-nav-small'
      ];

      const skeleton = await Promise.all(
        components
          .map((component) =>
          pfs.readFile(`${DIR_TEMPLATES}/${component}/template.mustache`, 'utf-8'))
      ).then((templates) => templates.reduce((skeleton, template, i) => {
        skeleton[components[i]] = template;
        return skeleton;
      }, {}));

      const skeletonLocalizations = pkg.langs
        .reduce((localizations, lang) => {
          localizations[lang] = components
            .reduce((localizedComponents, component) => {
              let i18n = `${DIR_TEMPLATES}/${component}/i18n/${lang}.json`;

              if (!pfs.existsSync(i18n)) {
                i18n = `${DIR_TEMPLATES}/${component}/i18n/en.json`;
              }

              if (!pfs.existsSync(i18n)) {
                log.warn(`Missing localization [${lang}] for ${component}.`)
              }

              return {
                ...localizedComponents,
                ...require(path.resolve(i18n))
              };
            }, {});
          return localizations;
        }, {});


      const promises = [];

      for (const domain of pkg.domains) {
        for (const site of sites) {
          const template = await pfs.readFile(`${DIR_TEMPLATES}/${site}/template.mustache`, 'utf-8');

          const vars = {
            ...skeletonLocalizations[domain.lang],
            ...require(path.resolve(`${DIR_TEMPLATES}/index/i18n/${domain.lang}.json`)),
            ...require(path.resolve(`${DIR_TEMPLATES}/${site}/i18n/${domain.lang}.json`)),
            site,
            domain: domain.domain,
            lang: domain.lang,
            year: (new Date()).getFullYear(),
            path: `${site === 'about' ? '/' : `/${site}`}`
          };

          for (const item of vars.menuItems) {
            item.active = item.id === site;
          }

          for (const item of vars.footerMenuItems) {
            item.active = item.id === site;
          };

          const destDir = `${DIR_DIST}/${domain.domain}/${domain.includeServer ? 'public' : ''}/${site !== 'about' ? `${site}/` : ''}`;

          await new Promise((resolve, reject) => {
            gulp.src(`${DIR_TEMPLATES}/index/template.mustache`)
              .pipe($.mustache(
                vars,
                {},
                {
                  view: template,
                  ...skeleton
                }
              ))
              .pipe($.rename((path) => {
                path.basename = 'index';
                path.extname = '.html';
                return path;
              }))
              .pipe(gulp.dest(destDir))
              .on('end', resolve)
              .on('error', reject);
          });
        }
      }

      done();
    }
  )
);

gulp.task('html',
  gulp.series(
    gulp.parallel(
      'templates',
      'css'
    ),
    () => Promise.all(
      pkg.domains
        .map((domain) => new Promise((resolve, reject) => {
          gulp.src(`${DIR_DIST}/${domain.domain}/**/*.html`)
            .pipe($.replace(DIRECTIVE_CSS_INCLUDE, (match) => `<style amp-custom>${ pfs.readFileSync(path.join(DIR_DIST, 'style.tmp.css'))}</style>`))
            .pipe($.replace(DIRECTIVE_INCLUCE_VERSION, (match) => `<meta name="version" content="${VERSION}" >`))
            .pipe($.replace(DIRECTIVE_INCLUDE_REVISED, (match => `<meta name="revised" content="${ DATE_NOW_ISO }"><meta name="date" content="${ DATE_NOW_ISO }">`)))
            .pipe($.embedSvg({
              root: path.join(__dirname, 'src/artwork'),
              createSpritesheet: true,
              xmlMode: false
            }))
            .pipe($.embedJson({
              root: path.join(DIR_SRC_ASSETS, 'json')
            }))
            .pipe($.htmlmin(OPTS_HTMLMIN))
            .pipe(gulp.dest(`${DIR_DIST}/${domain.domain}`))
            .on('end', resolve)
            .on('error', reject);
        })
      )
    )
  )
);

gulp.task('github:errorPages:prependFrontMatter',
  gulp.series(
    'html',
    () => Promise.all(
      pkg.domains
        .filter((domain) => domain.isGithub)
        .map((domain) => new Promise((resolve, reject) =>
           gulp.src(`${DIR_DIST}/${domain.domain}/404/index.html`)
            .pipe($.insert.prepend('---\npermalink: /404.html\n---\n'))
            .pipe($.rename((path) => {
              path.basename = '404';
              return path;
            }))
            .pipe(gulp.dest(`${DIR_DIST}/${domain.domain}`))
            .on('end', resolve)
            .on('error', reject)
        ))
    )
  )
);

gulp.task('github:errorPages',
  gulp.series(
    'github:errorPages:prependFrontMatter',
    () => Promise.all(
      pkg.domains
        .filter((domain) => domain.isGithub)
        .map((domain) => del(
          [
            `${DIR_DIST}/${domain.domain}/404/index.html`,
            `${DIR_DIST}/${domain.domain}/404`
          ],
          { force: true }
        ))
    )
  )
);

gulp.task('compile',
  gulp.series(
    'github:errorPages',
    () => del(
        path.join(DIR_DIST, 'style.tmp.css'),
        { force: true }
    ).then(() => {
      browserSync.reload();
      validateAMP();
    })
  )
);

gulp.task('assets',
  gulp.series(
    'clean:assets',
    () => Promise.all(
      pkg.domains
        .map((domain) => new Promise((resolve, reject) =>
          gulp.src(`${DIR_SRC}/assets/+(img|docs|poe)/*`)
          .pipe(gulp.dest(`${DIR_DIST}/${domain.domain}/${domain.includeServer ? 'public' : ''}/assets`))
            .on('end', resolve)
            .on('error', reject)
        ))
    )
  )
);

gulp.task('build',
  gulp.parallel(
    'compile',
    'assets',
    'sitemap',
    'seofiles',
    'server'
  )
);

gulp.task('watch',
  gulp.series(
    'build',
    (done) => {
      gulp.watch(`${DIR_SRC_SEOFILES}/**/*`, gulp.series('seofiles'));
      gulp.watch(FILES_SERVER, gulp.series('server'));
      gulp.watch(`${DIR_SRC}/**/*.+(css|mustache)`, gulp.series('compile'));
      gulp.watch(`${DIR_SRC_ASSETS}/*/*`, gulp.series('assets'));
      gulp.watch(`${DIR_SRC_ASSETS}/json/*`, gulp.series('compile'));
      gulp.watch(`${DIR_ARTWORK}/*.svg`, gulp.series('html'));
      done();
    }
  )
);

gulp.task('serve', (done) => {
  browserSync.init({
    server: {
      baseDir: `${DIR_DIST}/haensl.github.io`
    }
  });
  done();
  log.info('Production server listening on http://localhost:8080');
  log.info('Browser sync server listening on http://localhost:3000');
});

gulp.task('default', gulp.series('watch'));
